qiskit_aqua/utils/circuit_cache.py

```diff
@@ -0,0 +1,203 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018 IBM.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+""" A utility for caching and reparameterizing circuits, rather than compiling from scratch
+with each iteration. Note that if the circuit is transpiled aggressively such that rotation parameters
+cannot be easily mapped from the uncompiled to compiled circuit, caching will fail gracefully to
+standard compilation. This will be noted by multiple cache misses in the DEBUG log. It is generally safer to
+skip the transpiler (aqua_dict['backend']['skip_transpiler'] = True) when using caching.
+
+Caching is controlled via the aqua_dict['problem']['circuit_caching'] parameter. Caching naughty
+mode bypasses qobj validation before compiling and reuses the same qobj object over and over to
+avoid deepcopying. It is controlled via the aqua_dict['problem']['caching_reuse_qobj_skip_validation'] parameter.
+
+You may also specify a filename into which to store the cache as a pickle file, for circuits which
+are expensive to compile even the first time. The filename is set in aqua_dict['problem']['circuit_cache_file'].
+If a filename is present, the system will attempt to load from the file.
+
+In the event of an error, the system will fail gracefully, compile from scratch, and cache the new
+compiled qobj and mapping in the file location in pickled form. It will fail over 5 times before deciding
+that caching should be disabled."""
+
+import numpy as np
+import copy
+import pickle
+import logging
+
+from qiskit import QuantumRegister
+from qiskit.circuit import CompositeGate
+from qiskit.qobj import qobj_to_dict, Qobj
+
+from qiskit_aqua.aqua_error import AquaError
+
+logger = logging.getLogger(__name__)
+
+class CircuitCache:
+
+    def __init__(self,
+                 caching_naughty_mode = False,
```

Comment:
if we change the flag name in the declarative approach, do you think that we could align the name in the constructor?

Reply:
Yes - I'll rename everything when I split this into two features.